File: k-diff-pairs-in-an-array.java
// Time Complexity : O(n)
// Space Complexity : O(n)
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this : yes


// Your code here along with comments explaining your approach
//Hash map is created with frequency map, iterated over the hashmap to increment the counter based on the diff found.

class Solution {
    public int findPairs(int[] nums, int k) {
        int counter =0 ;
        Map<Integer,Integer> map = new HashMap<>();
        for(int i=0;i<nums.length;i++){
            map.put(nums[i],map.getOrDefault(nums[i],0)+1);
        }

        for(Map.Entry<Integer, Integer> entry:map.entrySet()){
            if(k>0 && map.containsKey(k+entry.getKey())){
                counter++;
            }else if(k==0 && entry.getValue() > 1){
                counter++;
            }
        }
        return counter;
    }
}

File: K-diff_Pairs_in_an_Array.py
from collections import Counter

# time complexity: O(n)
# space complexity: O(1)

class Solution:
    def findPairs(self, nums, k):
        if k < 0:
            return 0

        counter = Counter(nums)
        count = 0

        for num, frequency in counter.items():
            if (k == 0 and frequency > 1) or (k > 0 and num + k in counter):
                count += 1

        return count
